src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -30,19 +32,429 @@
+            Assert.Equal(textAndVersion.Version, actualVersion);
+
+            var newSolution2 = newSolution1.WithDocumentText(documentId, textAndVersion, PreservationMode.PreserveIdentity);
+            Assert.Same(newSolution1, newSolution2);
+
+            Assert.Throws<ArgumentNullException>(() => solution.WithDocumentText(documentId, (SourceText)null, PreservationMode.PreserveIdentity));
+            Assert.Throws<ArgumentOutOfRangeException>(() => solution.WithDocumentText(documentId, textAndVersion, (PreservationMode)(-1)));
+
+            Assert.Throws<InvalidOperationException>(() => solution.WithDocumentText((DocumentId)null, textAndVersion, PreservationMode.PreserveIdentity));
+            Assert.Throws<InvalidOperationException>(() => solution.WithDocumentText(s_documentId, textAndVersion, PreservationMode.PreserveIdentity));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        public void WithDocumentText_MultipleDocuments()
+        {
+            var solution = CreateSolutionWithProjectAndDocuments();
+            var documentId = solution.Projects.Single().DocumentIds.Single();
+            var text = SourceText.From("new text");
+
+            var newSolution1 = solution.WithDocumentText(new[] { documentId }, text, PreservationMode.PreserveIdentity);
+            Assert.True(newSolution1.GetDocument(documentId).TryGetText(out var actualText));
+            Assert.Equal(text, actualText);
+
+            var newSolution2 = newSolution1.WithDocumentText(new[] { documentId }, text, PreservationMode.PreserveIdentity);
+            Assert.Same(newSolution1, newSolution2);
+
+            // documents not in solution are skipped:
+            Assert.Same(solution, solution.WithDocumentText(new DocumentId[] { null }, text));
+            Assert.Same(solution, solution.WithDocumentText(new DocumentId[] { s_documentId }, text));
+
+            Assert.Throws<ArgumentNullException>(() => solution.WithDocumentText((DocumentId[])null, text, PreservationMode.PreserveIdentity));
+            Assert.Throws<ArgumentNullException>(() => solution.WithDocumentText(new[] { documentId }, null, PreservationMode.PreserveIdentity));
+            Assert.Throws<ArgumentOutOfRangeException>(() => solution.WithDocumentText(new[] { documentId }, text, (PreservationMode)(-1)));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        public void WithAdditionalDocumentText_SourceText()
+        {
+            var solution = CreateSolutionWithProjectAndDocuments();
+            var documentId = solution.Projects.Single().AdditionalDocumentIds.Single();
+            var text = SourceText.From("new text");
+
+            var newSolution1 = solution.WithAdditionalDocumentText(documentId, text, PreservationMode.PreserveIdentity);
+            Assert.True(newSolution1.GetAdditionalDocument(documentId).TryGetText(out var actualText));
+            Assert.Equal(text, actualText);
```

Comment:
```suggestion
            Assert.Same(text, actualText);
```

Since PreserveIdentity is being used.